Add a comment indicating the context under which a page was downloaded

Alch-Emi opened the pull request:
Currently, monolith leaves very little trace of where a page came from, or any other record of the conditions under which a page was saved.

It's very likely that, after a page is saved, a user needs to recall where that page originally came from, possibly because they didn't save a complicated page URL.  They might also need other information about how the page was recorded, such as the date the page was saved (useful if you're trying to determine if/when a page changed, or trying to figure out why a page and its monolith differ) and the version of monolith that was used to save it (useful if you're trying to decide whether or not to redownload a page with an updated version of monolith, for example).

This pull request prepends a short comment to the beginning of each downloaded page noting the page URL, the current date, and the version of monolith.  For example:
```html
<!--- Downloaded from https://example.com/ on Wed, 08 Jan 2020 19:22:30 using monolith v2.1.1 -->
```
This comment is added to all pages by default, but in the event that a user wants their page modified only minimally, this option can be disable via the `-C/--no-comment` option, which applies monolith's current behavior.

This change adds another direct dependency (`time-0.1.42`) in order to render the timestamp.  However, I don't think this is itself worth avoiding, since `time` is already a requirement of several other dependencies.  Because of this, the release binary only grows by ~14.8 kB (0.15%).

snshn commented:
This is a great addition! I remember IE6 doing something like that back in the day so it did cross my mind implementing it for monolith, too, at some point. The only addition I would make is a timezone (or specify the time is UTC). Showing the timezone other than UTC may compromise the location from which the document was saved, hence my vote goes for UTC. Using an ISO format could still be human readable yet provide ability to parse it easier to e.g. let some daemon update local version of a remote resource if the page ends up being outdated

Nits:
- the URL should probably go through a filter of sorts, e.g. remove the #fragment and any user:pass@ from it for security/privacy purposes
- when monolith is able to work with local files as well, we'll need to either exclude this comment completely or reword the "Downloaded" in this PR; could be "Compiled" for local and "Downloaded" for remote, if you think that having that comment there may be needed when a local document is saved.
- we could also consider appending the same timestamp to the filename in the future, e.g. when the name is meant to be auto-determined from the title of the page, (e.g. `-o -` or something of that sort)

snshn commented:
Having second thoughts about "metadata", I'll look for similar options in other popular tools and try to come up with better wording - both for the readme and --help.

Alch-Emi commented:
Alrighty!

I've switched it over to ISO8601 UTC, and cleaned up the URL beforehand.  Re: bullet # 2, this should be pretty trivial.  The current framework sets the comment by an `Option<&str>`, so just passing in something else switched based on the URL protocol should be pretty easy.

On bullet 3, that sounds like a really good idea.  Perhaps something like youtube-dl's syntax of putting formatting tokens in the name for substitution?  e.g. `monolith "https://example.com" -o "example - %d.html"` in order to save a file called `example - 2020-01-10.html`.  Mind if I open an issue to discuss it?

Also, would you mind elaborating on your last comment about metadata?

snshn commented on src/main.rs:
```diff
@@ -71,6 +72,7 @@ fn main() {
             app_args.silent,
         )
         .unwrap();
+        let downloaded_time = time::now_utc();
```
I think naming this variable `download_time` would be slightly more semantic

snshn commented on src/main.rs:
```diff
@@ -94,6 +96,25 @@ fn main() {
         );
 
+        if !app_args.no_context {
+            // Safe to unwrap: We just put this through an HTTP request
```
I'd change it to `// Safe to unwrap - put this through an HTTP request`

snshn commented on src/main.rs:
```diff
@@ -94,6 +96,25 @@ fn main() {
+            // Safe to unwrap: We just put this through an HTTP request
+            let mut clean_url = Url::parse(&final_url).unwrap();
+            clean_url.set_fragment(None);
+            // Safe to unwrap: must have a protocol and thus base 'cause we just used it.
```
to match the style of the previous comment, I think that `// Safe to unwrap - must have a protocol and thus base because we just used it` would be more fitting

snshn commented on src/main.rs:
```diff
@@ -94,6 +96,25 @@ fn main() {
+            html.insert_str(
+                0,
+                &format!(
+                    "<!--- Downloaded from {} on {} using {} v{} -->\n",
```
I believe HTML comments don't need more than two dashes in the beginning (https://html.com/tags/comment-tag/)

snshn commented:
> On bullet 3, that sounds like a really good idea. Maybe something like youtube-dl's syntax of putting formatting tokens in the name for substitution? e.g. `monolith "https://example.com" -o "example - %d.html"` in order to save a file called `example - 2020-01-10.html`. Mind if I open an issue to discuss it?

I wasn't aware youtube-dl had such capability. I now wonder if there's some sort of standard filename tokens. It looks like curl has a slightly different (similar to regex) approach https://unix.stackexchange.com/questions/91570/write-output-of-wget-or-curl-to-a-custom-filename-based-on-the-url
Either way, it's outside of the scope of this PR.

> Also, would you mind elaborating on your last comment about metadata?

I was thinking that that comment tag may be used to store more data in the future, e.g. things like 
```html
<!--
    Downloaded from ... v2.1.2
    File size (original & final): 5.6MB/2.8MB
    Failed to retrieve: (a list of resources that responded with 401/403/404)
   (etc)
-->
```
and wasn't sure what term other than "metadata" to use for the CLI flag in such a case. I guess `--no-context` is broad enough to not make it misleading no matter what other data monolith may append to that comment node in the future. In all honesty I'm still on the fence between `--no-comment` and `--no-context`; the description can be changed later, while renaming the flag itself is a breaking change.

snshn commented:
I believe flag options `-M/--no-metadata` would be more specific, as "context" seems to be too ambiguous, and `--no-comment` may confuse the user into thinking its purpose is to remove all comments from the document; besides the tag will likely hold more content in the future as the program obtains more features.
Sorry about holding this for so long. If you could please resolve the conflict and okay with changing it to `-M/--no-metadata` I'll merge it into master right away.

rhysd commented:
I have created a small PR to merge upstream `master` into this branch:

https://github.com/Alch-Emi/monolith/pull/1

@Alch-Emi Please review it and merge it if it looks good. Merging the PR will resolve all conflicts in `main.rs`.
